keps/sig-api-machinery/5073-declarative-validation-with-validation-gen/README.md

```diff
@@ -0,0 +1,1246 @@
+                        verifyVersionedValidationEquivalence(t, obj, nil)
+
+                        old, err := legacyscheme.Scheme.New(gv.WithKind(kind))
+                        if err != nil {
+                                t.Fatalf("could not create a %v: %s", kind, err)
+                        }
+                        f.Fuzz(old)
+                        verifyVersionedValidationEquivalence(t, obj, old)
+                }
+        }
+}
+```
+
+###### strategy_test.go vs validation_test.go
+
+Currently, validation logic and associated tests are logically split across `validation.go` and `strategy.go`. The hand-written validation functions and associated tests reside in `validation.go` and `validation_test.go`  while `strategy.go` determines when to invoke these functions during API object creation and updates. With the introduction of declarative validation (controlled by the `DeclarativeValidation` feature gate)  the current logic split is worth considering moving from `validation_test.go` to `strategy_test.go` as the current logic split in the test is unfavorable for the migration as it requires additional plumbing work.
+
+The current approach in the prototype experiment to migrate ReplicationController involves directly injecting declarative validation and equivalency tests into `validation_test.go`. This is achieved by conditionally appending calls to `rest.ValidateDeclaratively` within the existing test cases, based on the `DeclarativeValidation` feature gate's status. This allows for a direct comparison of the outputs between hand-written and declarative validation within the same test framework.
+
+**Moving the feature gate check and declarative validation logic to <code>strategy_test.go</code> could offer several advantages</strong>:
+
+*   **Reduced Test Duplication:** `validation_test.go` could be simplified, as it would no longer need to handle both hand-written and declarative validation paths.
+*   **Clearer Separation of Concerns:** `strategy.go` would be responsible for determining _when_ to validate, while `validation.go` would handle _how_ to validate.
+*   **Easier Migration:**  Transitioning to a fully declarative model would be smoother, as the core validation logic would already be invoked through the strategy.
+
+However, this approach also has potential drawbacks:
+
+*   **Additional PRs for Initial Migration:** Moving existing hand-written validation logic from `validation.go` to `strategy.go` would result in additional work via a PR to move the related code but it would be straightforward (only moving files).
+
+Given the low complexity of moving this code prior to the changes, the enhanced logic split of moving the code, and the reduced work for the migration that moving this code would have currently the plan for Declarative Validation is that the current `validation_test.go` tests are moved to `strategy_test.go` in a PR prior to the migration PR.  
+
+###### Error Message Equivalence
```

Comment:
Could we simply change the hand written message to match in most cases?

Reply:
One of the slippery-slope problems with this KEP is that it demands we look at the error strings, and the desire to improve those (e.g. not telling a user that their input has to match a 50-character-long regex) is strong.  As we proceed, I expect some changes will occur just as a result of howe the declarative rules are factored.

We should try to keep both paths in sync, but the strings themselves might change from today (hopefully only better)